sdk/core/azure-core/tests/test_universal_pipeline.py

```diff
@@ -49,9 +50,105 @@
+
+
+def test_telemetry_policy_keep_existing_user_agent():
+    url = "https://bing.com"
+    config = Configuration()
+    config.headers_policy = HeadersPolicy({'User-Agent': 'ExistingUserAgentValue'})
+    config.telemetry_policy = TelemetryPolicy("storage/1.01")
+
+    client = PipelineClient(base_url=url, config=config)
+    request = client.get(url)
+    pipeline_response = client._pipeline.run(
+        request,
+        user_agent="AzCopy/10.0.4-Preview"
+    )
+    request = pipeline_response.http_request
+    user_agent = request.headers.get('User-Agent')
+    assert user_agent.startswith('AzCopy/10.0.4-Preview azsdk-python-storage/1.01')
+    assert user_agent.find('ExistingUserAgentValue')
+
+
+def test_telemetry_policy_with_dynamic_telemetry():
+    url = "https://bing.com"
+    config = Configuration()
+    config.headers_policy = HeadersPolicy()
+    config.telemetry_policy = TelemetryPolicy("storage/1.01")
+
+    client = PipelineClient(base_url=url, config=config)
+    request = client.get(url)
+    pipeline_response = client._pipeline.run(
+        request,
+        user_agent="AzCopy/10.0.4-Preview",
+        telemetry="class=BlobClient;method=DownloadFile;blobType=Block"
+    )
+
+    request = pipeline_response.http_request
+    user_agent = request.headers.get('User-Agent')
+    telemetry = request.headers.get('X-MS-AZSDK-Telemetry')
+    assert user_agent.startswith('AzCopy/10.0.4-Preview azsdk-python-storage/1.01')
+    assert telemetry == 'class=BlobClient;method=DownloadFile;blobType=Block'
+
+
+def test_telemetry_policy_with_telemetry_disabled():
+    with mock.patch.dict('os.environ', {'AZURE_TELEMETRY_DISABLED': "True"}):
+        policy = TelemetryPolicy('storage/1.01')
+
+        request = HttpRequest('GET', 'https://bing.com')
+        policy.on_request(PipelineRequest(request, PipelineContext(None)))
+        assert request.headers.get("user-agent", None) is None
+
+
+def test_telemetry_policy_with_overwrite():
+    url = "https://bing.com"
+    config = Configuration()
+    config.headers_policy = HeadersPolicy({'User-Agent': 'ExistingUserAgentValue'})
+    config.telemetry_policy = TelemetryPolicy("storage/1.01")
+
+    client = PipelineClient(base_url=url, config=config)
+    request = client.get(url)
+    pipeline_response = client._pipeline.run(
+        request,
+        user_agent="OverwritingUserAgent",
+        user_agent_overwrite=True
+    )
+
+    request = pipeline_response.http_request
+    user_agent = request.headers.get('User-Agent')
+    assert user_agent == 'OverwritingUserAgent'
+
+
+def test_telemetry_policy_from_environment():
+    with mock.patch.dict('os.environ', {'AZURE_HTTP_USER_AGENT': "UserAgentFromEnv"}):
+        policy = TelemetryPolicy(user_agent_use_env=True)
+
+        request = HttpRequest('GET', 'https://bing.com')
+        policy.on_request(PipelineRequest(request, PipelineContext(None)))
+        assert request.headers["user-agent"].endswith("UserAgentFromEnv")
+
+
```

Comment:
Here too I think there are more cases that could be covered. E.g. verify the correct behavior when `user_agent_use_env=True` *and* a `user_agent` option is passed.  I think a *full* matrix of combinations is:

* `user_agent_use_env` = True/False
* `AZURE_HTTP_USER_AGENT` set/not set
* `user_agent` option passed/not passed
* `overwrite` = True/False
* `AZURE_TELEMETRY_DISABLED` enabled/disabled
* `telemetry` passed/not passed
* has/does not have existing user agent

I'm not saying all 128 of those should be explicitly covered :) just that I think there are more that are potentially interesting to maintain under test than are currently here.

Reply:
Added several more tests to try to cover the more interesting cases.